star-aligner/src/lib.rs

```diff
@@ -109,92 +157,211 @@ impl Drop for StarAligner {
 }
 
 unsafe impl Send for StarAligner {}
+unsafe impl Sync for StarAligner {}
+
+impl Clone for StarAligner {
+    fn clone(&self) -> StarAligner {
+        StarAligner::new(self.index.clone())
+    }
+}
 
 impl StarAligner {
     pub fn new(index: StarIndex) -> Self {
         let aligner = unsafe { star_sys::init_aligner_from_ref(index.0.as_ref().index) };
         Self {
             aligner,
             index,
-            sam_buf: Vec::new(),
-            aln_buf: Vec::new(),
-            fastq1: Vec::new(),
-            fastq2: Vec::new(),
+            // sam_buf: SamRecord::default(),
+            // fastq1: FastqRecord::default(),
+            // fastq2: FastqRecord::default(),
         }
     }
 
     pub fn get_header(&self) -> &Header {
         &self.index.0.header
     }
 
-    fn align_read(&mut self, name: &[u8], seq: &[u8], qual: &[u8]) -> Vec<sam::Record> {
-        Self::prepare_fastq(&mut self.fastq1, name, seq, qual);
+    // align single read in parallel
+    pub fn align_reads(
+        &mut self,
+        fqrecs: &mut [fastq::Record],
+        n_threads: usize,
+    ) -> Result<Vec<SamRecord>> {
+        // set thread pool
+        rayon::ThreadPoolBuilder::new()
+            .num_threads(n_threads)
+            .build_global()?;
+        // split the reads into chunks by the number of threads
+        let chunk_size = (fqrecs.len() + n_threads - 1) / n_threads;
+        let fqrecs_chunks: Vec<_> = fqrecs.chunks(chunk_size).collect();
+
+        // construct an Arc mut clone to share aligner among threads
+        let aligner = Arc::new(Mutex::new(self.clone()));
+
+        // align reads in parallel
+        let records: Vec<SamRecord> = fqrecs_chunks
+            .par_iter()
+            .flat_map(|chunk| {
+                chunk
+                    .iter()
+                    .flat_map(|fqrec| {
+                        // lock the aligner
+                        let mut aligner = aligner.lock().unwrap();
```

Comment:
Using `Mutex` means there will be only one single thread is holding the aligner and doing the job at a time. You're not using multithreading. You should simply clone the aligner within "flat_map" loop.